### ISSUE TITLE
Gradient is incorrect for log pdf of Normal distribution

### ISSUE THREAD

**Issue by altosaar** [tensorflow/tensorflow#7236]
Am I missing something basic here?

Gradients of the log pdf of normal distributions are zero. I might be going crazy, it is late at night.

```
In [12]: mu = tf.get_variable('mu', [], 'float32')
In [13]: q = dist.Normal(mu, 1.)
In [14]: log_q = q.log_pdf(q.sample())
In [15]: grad = tf.gradients(log_q, mu)
In [16]: sess.run(tf.global_variables_initializer())
In [17]: sess.run(grad)
Out[17]: [0.0]
In [18]: tf.__version__
Out[18]: '0.12.1'
```

**Comment by altosaar** [comment]
Thanks to @poolio - it's because of reparameterization:

```
eps ~ N(0, 1)
z = mu + sigma * eps
q.log_pdf(z) ~= -(z - mu)^2 / 2 = - (mu + sigma * eps - mu)^2 / 2 = - (sigma * eps)^2 / 2
```

so it makes sense that the gradient is zero. 

This is very unexpected behavior and broke my regression model. 

Distributions should abstract parameters: a sample should be treated as a new tensor. 

Is there a discussion of this anywhere?

**Comment by dustinvtran** [comment]
there is some in the reparameterization type doc. generally when calculating score functions, you have to perform `tf.gradients(q.log_prob(tf.stop_gradients(z)), mu)`. i agree having to be aware of how the sampling mechanism is implemented can be unintuitive. both use cases—taking gradients through samples or not through samples—are useful though.

**Comment by altosaar** [comment]
Thanks Dustin! 

I think it's weird to have to remember this.

For example, if you have a Bernoulli distribution, you don't need the tf.stop_gradients. 

Samples from distributions behaving differently in various contexts adds a lot of complexity to my code.

**Comment by yaroslavvb** [comment]
so is this expected behavior, or is it a bug in TensorFlow?

**Comment by jvdillon** [comment]
On Fri, Feb 3, 2017 at 1:21 PM, Yaroslav Bulatov <notifications@github.com>
wrote:

> so is this expected behavior, or is it a bug in TensorFlow?
>

Arguably yes/yes, yes/no, no/yes, no/no. :)

Many of the samplers don't have analytic gradients due to being
dynamic-loop based, eg rejection samplers.  We have a reparameterization
annotation for each class.  However, we should consider at least triggering
a warning if backpropping into a non-reparameterizable sampler.  Noteworthy
example: gradient of Gamma.sample is None wrt to shape param but -x/scale
wrt to scale param (where x~Gamma).

**Comment by yaroslavvb** [comment]
ok, I'll go the least effort route and interpret it as "working as intended" :) If this is a bug/feature request, feel free to provide more concrete details of what should be changed, and I'll reopen
